doc/developer/design/20230216_role_based_access_control.md

```diff
@@ -0,0 +1,384 @@
+    - Requires membership of `<new_owner>`.
+    - Requires `CREATE` privilege on the schema where `<object_name>` resides if the object resides in a schema.
+        - Rationale is that this is equivalent to `DROP` then `CREATE`.
+    - Requires `CREATE` privilege on the database where `<object_name>` resides if the object is a database.
+- `REASSIGN OWNED BY <old_role> TO <new_role>`
+    - Transfers ownership of all objects owned by `<old_role>` to `<new_role>`.
+    - Can only be run by a member of `<old_role>` and `<new_role>` or a superuser.
+    - Requires `CREATE` privilege on all schemas and databases where all objects reside.
+        - TODO: This isn't explicitly stated in th docs, but would make sense base on the `ALTER` privileges. I will
+          double-check this.
+    - In PostgreSQL, this only affects the current database. We will diverge here and have it affect all databases.
+- `DROP OWNED BY <name> [ CASCADE | RESTRICT]`
+    - Drops all objects owned by `<name>`.
+    - Requires membership of `<name>`.
+    - In PostgreSQL, this only affects the current database. We will diverge here and have it affect all databases.
+    - Revokes all privileges granted to `<name>`.
+    - Does not drop owned databases or clusters.
+    - Default is `RESTRICT`.
+
+We will update `DROP ROLE` so that roles cannot be dropped until it meets the following criteria:
+
+- No objects are owned by the role.
+- The role contains no privileges.
+
+We will update `DROP <object>` so that it revokes all privileges on `<object>`.
+
+#### Implementation Details
+
+- Privileges will be stored in the catalog.
+- Privileges will be checked before operations in the sequencer.
+
+#### Out of Scope for Project
+
+- `CURRENT_ROLE`, `CURRENT_USER`, `SESSION_USER`, aliases in `GRANT`, `REVOKE`, `ALTER`, `REASSIGN OWNED`
+  and `DROP OWNED`.
+- `GRANTED BY` option for `GRANT` and `REVOKE`.
+- `WITH GRANT OPTION` in `GRANT`.
+- `GRANT OPTION FOR` in `REVOKE`.
+- The following privileges:
+    - `TRUNCATE`
+    - `REFERENCES`
+    - `TRIGGER`
+    - `CONNECT`
+    - `TEMPORARY`
+    - `EXECUTE`
+    - `SET`
+    - `ALTER SYSTEM`
+- The following object types:
+    - `DOMAIN`
+    - `FUNCTION` or `PROCEDURE`
+    - `FOREIGN DATA WRAPPER`
+    - `FOREIGN SERVER`
+    - `LANGUAGE`
+    - `LARGE OBJECT`
+    - `PARAMETER`
+    - `SEQUENCE`
+    - `Table column`
+    - `TABLESPACE`
+    - `TYPE`
+- Adding the necessary pg views to support all role based `psql` meta-commands.
+
+## Alternatives
+
+- PostgreSQL breaks from the SQL standard in a couple of places in their RBAC implementation. We may want to consider a
+  design that is more in line with the SQL standard than PostgreSQL.
+
+## Open questions
+
+- PostgreSQL grants certain privileges on certain object types to all roles. None of those overlap with the set of
+  objects and privileges we have, but do we want to do something similar?
```

Comment:
Here's the full set from https://www.postgresql.org/docs/current/ddl-priv.html

> the default privileges granted to PUBLIC are as follows: CONNECT and TEMPORARY (create temporary tables) privileges for databases; EXECUTE privilege for functions and procedures; and USAGE privilege for languages and data types (including domains). 

It goes on to say

> The object owner can, of course, REVOKE both default and expressly granted privileges. (For maximum security, issue the REVOKE in the same transaction that creates the object; then there is no window in which another user can use the object.) Also, these default privilege settings can be overridden using the [ALTER DEFAULT PRIVILEGES](https://www.postgresql.org/docs/current/sql-alterdefaultprivileges.html) command.

Reply:
> the default privileges granted to PUBLIC are as follows: CONNECT and TEMPORARY (create temporary tables) privileges for databases; EXECUTE privilege for functions and procedures; and USAGE privilege for languages and data types (including domains).

💡 

We _do_ support custom data types in a limited form! (See `CREATE TYPE`.) So we should probably follow PostgreSQL and default grant `USAGE` on data types to `PUBLIC`!

Reply:
I was planning on explicitly leaving any privileges on types out of this initial project because it didn't seem high priority. Though, I'm happy to add them in if you disagree.